server/src/main/java/io/crate/analyze/validator/SemanticSortValidator.java

```diff
@@ -94,13 +94,14 @@ public Void visitFunction(Function symbol, SortContext context) {
                                   symbol.valueType())
                 );
             }
+            boolean wasAlreadyInFunction = context.inFunction;
             try {
                 context.inFunction = true;
                 for (Symbol arg : symbol.arguments()) {
                     arg.accept(this, context);
                 }
             } finally {
-                context.inFunction = false;
+                context.inFunction = wasAlreadyInFunction;
             }
```

Comment:
Checked what happens in working scenario (order by 1). 
In this case we start with `visitSymbol` with Symbol being an Alias and inner Symbol being Function-> we are not in function, return type  == alias return type == function return type == boolean, so we just short cut there.

I think we can simplify a fix: instead of introducing another flag we can drop `context.inFunction = false `because that would reflect this PR: once in function, never reset the flag in the inner functions. If we  order by expression, function is a top level symbol(?) and `context.inFunction = false` supposed to be last in visitors chain and don't affect any other visit... calls.

Reply:
Thanks for reviewing @BaurzhanSakhariev .

> Checked what happens in working scenario (order by 1).
In this case we start with visitSymbol with Symbol being an Alias and inner Symbol being Function-> we are not in function, return type == alias return type == function return type == boolean, so we just short cut there.

This actually is a source of another bug, the short cut could allow MatchPredicate to pass through I think. Will follow up on this.

Reply:
>This actually is a source of another bug

I had a suspect that shortcut aliased function basically leads to skipping all logic we have in the visitFunction --> not throwing error when facing a type which doesn't support comparison. But I think we cannot hit any issue as we don't throw inside Function anyway.

Good point regarding MATCH, I haven't checked this. I think we can do it in another PR to not block this one.

Reply:
Another thought:

The whole logic of this validator is "do not throw inside Function" and fix is "do not throw when have nested functions", so I think we can remove almost all logic from `visitFunction` and make it
```
@Override
public Void visitFunction(Function symbol, SortContext context) {
     context.inFunction = true;
     return null;
}
```

I was even thinking about dropping `visitFunction` completely (I tried locally, analyzer tests pass) but we still need to know whether we are inside function so that we don't throw on `ORDER BY geoshape is null`.  geoshape doesn't suppor tcomparison but since it's inside IS NULL we allow ORDER BY. 

Looks like we don't have enough tests to cover that "we cannot drop visitFunction"  but that's a different topic.

Reply:
We need to validate the return type of the function. Also need to visit the arguments to catch `match`es.

Reply:
To elaborate on `don't process Function args` and don't need `symbol instanceof Function`

If we call `visitFunction` coming to `some_func1`, it means that `ORDER BY` Symbol is also a Function 
(maybe `some_func1` or another one).

We can have standalone non-function symbols, next to `some_func1` but we need some operators to connect them into an expression, last applied operator would be some_func2 --> Symbol of the ORDER BY -> symbol you referred in `symbol instanceof Function`

Reply:
>Also need to visit the arguments to catch matches.

Ah right, so match inside scalar is not allowed? Do we have tests for that?

If so, I think we can keep processing args but don't need new boolean argument `symbol instanceof Function`